File: AI_Challenges/AI_Challenges_TN/challenge-10/rules.py
"""8 fraud detection rules. Each returns a list of RuleFlag dicts."""
import json
import pandas as pd

AUTO_THRESHOLD = 50_000
CLUSTER_LOW = 47_500

SURGICAL_PREFIX = 'S-'

BUNDLES = {
    'BUNDLE-01': frozenset(['P-0100', 'P-0101', 'P-0102']),
    'BUNDLE-02': frozenset(['P-0103', 'P-0104', 'P-0105']),
    'BUNDLE-03': frozenset(['P-0106', 'P-0107', 'P-0108']),
    'BUNDLE-04': frozenset(['P-0109', 'P-0110', 'P-0111']),
    'BUNDLE-05': frozenset(['P-0112', 'P-0113', 'P-0114']),
}

VALID_DX_PROC = {
    'J06.9':   {'P-0100','P-0101','P-0102'},
    'K21.0':   {'P-0103','P-0104'},
    'M54.5':   {'P-0105','P-0106','S-0300'},
    'J45.909': {'P-0107','P-0108'},
    'I10':     {'P-0109','P-0110'},
    'E11.9':   {'P-0111','P-0112'},
    'K29.70':  {'P-0113','P-0114'},
    'N39.0':   {'P-0115','P-0116'},
    'J18.9':   {'P-0117','P-0118','S-0301'},
    'M79.3':   {'P-0119','P-0120','S-0302'},
}


def _parse_procs(val) -> list:
    if isinstance(val, list):
        return val
    try:
        return json.loads(val)
    except Exception:
        return [val] if val else []


# ── Rule 1: Duplicate claim ───────────────────────────────────────────────────
def rule_duplicate(df: pd.DataFrame) -> dict:
    key = ['member_id', 'provider_id', 'claim_date', 'diagnosis_code']
    counts = df.groupby(key)['claim_id'].transform('count')
    flagged = {}
    for _, row in df[counts > 1].iterrows():
        flagged[row['claim_id']] = {
            'rule': 'duplicate_claim', 'severity': 5,
            'evidence': (f"Duplicate: member {row['member_id']} + provider {row['provider_id']} "
                         f"+ date {row['claim_date']} + diagnosis {row['diagnosis_code']} appears multiple times"),
        }
    return flagged


# ── Rule 2: Rapid re-submission ───────────────────────────────────────────────
def rule_rapid_resubmission(df: pd.DataFrame) -> dict:
    flagged = {}
    df2 = df[['claim_id','member_id','diagnosis_code','claim_date']].copy()
    df2['claim_date'] = pd.to_datetime(df2['claim_date'])
    df2 = df2.sort_values(['member_id','diagnosis_code','claim_date'])
    df2['prev_date'] = df2.groupby(['member_id','diagnosis_code'])['claim_date'].shift(1)
    df2['days_since'] = (df2['claim_date'] - df2['prev_date']).dt.days
    for _, row in df2[df2['days_since'].between(1, 6, inclusive='both')].iterrows():
        flagged[row['claim_id']] = {
            'rule': 'rapid_resubmission', 'severity': 3,
            'evidence': (f"Same member {row['member_id']} + diagnosis {row['diagnosis_code']} "
                         f"resubmitted {int(row['days_since'])} day(s) after previous claim"),
        }
    return flagged


# ── Rule 3: Upcoding ─────────────────────────────────────────────────────────
def rule_upcoding(df: pd.DataFrame) -> dict:
    flagged = {}
    df2 = df.copy()
    df2['_procs'] = df2['procedure_codes'].apply(_parse_procs)
    df2['_primary_proc'] = df2['_procs'].apply(lambda p: p[0] if p else None)
    stats = df2.groupby('_primary_proc')['submitted_amount'].agg(['mean','std','count'])
    stats = stats[stats['count'] >= 5]
    merged = df2.merge(stats, left_on='_primary_proc', right_index=True, how='left')
    for _, row in merged.dropna(subset=['mean','std']).iterrows():
        if row['std'] > 0:
            z = (row['submitted_amount'] - row['mean']) / row['std']
            if z > 2:
                flagged[row['claim_id']] = {
                    'rule': 'upcoding', 'severity': 4,
                    'evidence': (f"Amount {row['submitted_amount']:,.0f} for procedure {row['_primary_proc']} "
                                 f"is {z:.1f} std deviations above mean {row['mean']:,.0f}"),
                }
    return flagged


# ── Rule 4: Unbundling ────────────────────────────────────────────────────────
def rule_unbundling(df: pd.DataFrame) -> dict:
    flagged = {}
    for _, row in df.iterrows():
        procs = frozenset(_parse_procs(row['procedure_codes']))
        for bundle_name, components in BUNDLES.items():
            if components.issubset(procs):
                flagged[row['claim_id']] = {
                    'rule': 'unbundling', 'severity': 3,
                    'evidence': (f"Procedures {sorted(components)} are individual components of "
                                 f"{bundle_name} and should be billed as a single bundled code"),
                }
                break
    return flagged


# ── Rule 5: Phantom billing ───────────────────────────────────────────────────
def rule_phantom_billing(df: pd.DataFrame) -> dict:
    flagged = {}
    daily = df.groupby(['provider_id','claim_date'])['claim_id'].transform('count')
    for _, row in df[daily > 30].iterrows():
        cnt = daily.loc[row.name]
        flagged[row['claim_id']] = {
            'rule': 'phantom_billing', 'severity': 5,
            'evidence': (f"Provider {row['provider_id']} ({row['provider_name']}) submitted "
                         f"{cnt} claims on {row['claim_date']} (threshold: 30)"),
        }
    return flagged


# ── Rule 6: Weekend anomaly ───────────────────────────────────────────────────
def rule_weekend_anomaly(df: pd.DataFrame) -> dict:
    flagged = {}
    df2 = df.copy()
    df2['_procs'] = df2['procedure_codes'].apply(_parse_procs)
    df2['_is_surgical'] = df2['_procs'].apply(lambda ps: any(p.startswith(SURGICAL_PREFIX) for p in ps))

    total_by_prov = df2.groupby('provider_id').size()
    wknd_by_prov  = df2[df2['is_weekend']].groupby('provider_id').size()
    wknd_rate = (wknd_by_prov / total_by_prov).fillna(0)

    for _, row in df2[df2['is_weekend'] & df2['_is_surgical']].iterrows():
        rate = wknd_rate.get(row['provider_id'], 0)
        if rate < 0.05:
            flagged[row['claim_id']] = {
                'rule': 'weekend_anomaly', 'severity': 4,
                'evidence': (f"Surgical procedure on weekend for provider {row['provider_id']} "
                             f"whose historical weekend rate is {rate*100:.1f}% (threshold: 5%)"),
            }
    return flagged


# ── Rule 7: Diagnosis-procedure mismatch ─────────────────────────────────────
def rule_dx_proc_mismatch(df: pd.DataFrame) -> dict:
    flagged = {}
    for _, row in df.iterrows():
        dx = row['diagnosis_code']
        if dx not in VALID_DX_PROC:
            continue
        valid = VALID_DX_PROC[dx]
        procs = _parse_procs(row['procedure_codes'])
        mismatched = [p for p in procs if p not in valid]
        if mismatched:
            flagged[row['claim_id']] = {
                'rule': 'dx_proc_mismatch', 'severity': 4,
                'evidence': (f"Procedure(s) {mismatched} not clinically associated with "
                             f"diagnosis {dx} (valid: {sorted(valid)})"),
            }
    return flagged


# ── Rule 8: Amount clustering ─────────────────────────────────────────────────
def rule_amount_clustering(df: pd.DataFrame, threshold: float = AUTO_THRESHOLD) -> dict:
    flagged = {}
    low = threshold * 0.95
    for _, row in df[(df['submitted_amount'] >= low) & (df['submitted_amount'] < threshold)].iterrows():
        pct = row['submitted_amount'] / threshold * 100
        flagged[row['claim_id']] = {
            'rule': 'amount_clustering', 'severity': 2,
            'evidence': (f"Amount {row['submitted_amount']:,.0f} is {pct:.1f}% of the "
                         f"{threshold:,.0f} auto-approval threshold (range: {low:,.0f}–{threshold:,.0f})"),
        }
    return flagged


# ── Run all rules ─────────────────────────────────────────────────────────────
def run_all_rules(df: pd.DataFrame) -> dict[str, list]:
    """Returns {claim_id: [flag, ...]}"""
    all_flags: dict[str, list] = {cid: [] for cid in df['claim_id']}

    for rule_fn in [
        rule_duplicate, rule_rapid_resubmission, rule_upcoding,
        rule_unbundling, rule_phantom_billing, rule_weekend_anomaly,
        rule_dx_proc_mismatch, rule_amount_clustering,
    ]:
        for cid, flag in rule_fn(df).items():
            all_flags[cid].append(flag)

    return all_flags

File: AI_Challenges/AI_Challenges_TN/challenge-10/src/rules.ts
import type { Claim, RuleFlag } from './types.js'
import {
  SEVERITY, BUNDLES, VALID_DX_PROCS, CLUSTER_LOW, CLUSTER_HIGH,
  isSurgical, WEEKEND_RATE_THRESHOLD, PHANTOM_THRESHOLD,
  MIN_PROC_SAMPLES, RAPID_WINDOW_DAYS,
} from './constants.js'

function flag(claim_id: string, rule: string, triggered: boolean, evidence: string): RuleFlag {
  return { claim_id, rule, triggered, severity: SEVERITY[rule] ?? 1, evidence }
}

// Rule 1: Duplicate claim — same member, provider, date, diagnosis
export function ruleDuplicate(claims: Claim[]): RuleFlag[] {
  const counts = new Map<string, string[]>()
  for (const c of claims) {
    const key = `${c.member_id}|${c.provider_id}|${c.claim_date}|${c.diagnosis_code}`
    const arr = counts.get(key) ?? []
    arr.push(c.claim_id)
    counts.set(key, arr)
  }
  return claims.map(c => {
    const key = `${c.member_id}|${c.provider_id}|${c.claim_date}|${c.diagnosis_code}`
    const group = counts.get(key) ?? []
    const triggered = group.length > 1
    return flag(c.claim_id, 'duplicate', triggered,
      triggered ? `Duplicate group: ${group.join(', ')}` : '')
  })
}

// Rule 2: Rapid re-submission — same member+diagnosis within 7 days
export function ruleRapidResubmission(claims: Claim[]): RuleFlag[] {
  // Group by member+diagnosis, sort by date
  const groups = new Map<string, Claim[]>()
  for (const c of claims) {
    const key = `${c.member_id}|${c.diagnosis_code}`
    const arr = groups.get(key) ?? []
    arr.push(c)
    groups.set(key, arr)
  }
  for (const arr of groups.values()) arr.sort((a, b) => a.claim_date.localeCompare(b.claim_date))

  const triggered = new Map<string, string>()
  for (const arr of groups.values()) {
    for (let i = 1; i < arr.length; i++) {
      const prev = new Date(arr[i - 1].claim_date)
      const curr = new Date(arr[i].claim_date)
      const daysDiff = (curr.getTime() - prev.getTime()) / 86_400_000
      if (daysDiff <= RAPID_WINDOW_DAYS) {
        triggered.set(arr[i].claim_id,
          `Re-submission ${daysDiff.toFixed(0)}d after ${arr[i-1].claim_id} (same member+diagnosis)`)
      }
    }
  }
  return claims.map(c => flag(c.claim_id, 'rapid_resubmission', triggered.has(c.claim_id), triggered.get(c.claim_id) ?? ''))
}

// Rule 3: Upcoding — amount > mean + 2*std per procedure_code
export function ruleUpcoding(claims: Claim[]): RuleFlag[] {
  // Build stats per procedure code
  const procAmounts = new Map<string, number[]>()
  for (const c of claims) {
    for (const proc of c.procedure_codes) {
      const arr = procAmounts.get(proc) ?? []
      arr.push(c.submitted_amount)
      procAmounts.set(proc, arr)
    }
  }
  const procStats = new Map<string, { mean: number; std: number }>()
  for (const [proc, amounts] of procAmounts) {
    if (amounts.length < MIN_PROC_SAMPLES) continue
    const mean = amounts.reduce((s, v) => s + v, 0) / amounts.length
    const variance = amounts.reduce((s, v) => s + (v - mean) ** 2, 0) / amounts.length
    procStats.set(proc, { mean, std: Math.sqrt(variance) })
  }

  return claims.map(c => {
    for (const proc of c.procedure_codes) {
      const stats = procStats.get(proc)
      if (!stats) continue
      const threshold = stats.mean + 2 * stats.std
      if (c.submitted_amount > threshold) {
        return flag(c.claim_id, 'upcoding', true,
          `Amount $${c.submitted_amount.toFixed(2)} > mean+2σ ($${threshold.toFixed(2)}) for procedure ${proc}`)
      }
    }
    return flag(c.claim_id, 'upcoding', false, '')
  })
}

// Rule 4: Unbundling — claim has ALL individual codes for a known bundle
export function ruleUnbundling(claims: Claim[]): RuleFlag[] {
  return claims.map(c => {
    const procs = new Set(c.procedure_codes)
    for (const [bundleName, components] of Object.entries(BUNDLES)) {
      if (components.every(code => procs.has(code))) {
        return flag(c.claim_id, 'unbundling', true,
          `Contains all components of ${bundleName}: ${components.join(', ')}`)
      }
    }
    return flag(c.claim_id, 'unbundling', false, '')
  })
}

// Rule 5: Phantom billing — provider with > 30 claims on a single day
export function rulePhantomBilling(claims: Claim[]): RuleFlag[] {
  const dailyCounts = new Map<string, string[]>()
  for (const c of claims) {
    const key = `${c.provider_id}|${c.claim_date}`
    const arr = dailyCounts.get(key) ?? []
    arr.push(c.claim_id)
    dailyCounts.set(key, arr)
  }
  return claims.map(c => {
    const key = `${c.provider_id}|${c.claim_date}`
    const count = (dailyCounts.get(key) ?? []).length
    const triggered = count > PHANTOM_THRESHOLD
    return flag(c.claim_id, 'phantom_billing', triggered,
      triggered ? `Provider ${c.provider_id} submitted ${count} claims on ${c.claim_date}` : '')
  })
}

// Rule 6: Weekend anomaly — surgical procedure on weekend, provider weekend rate < 5%
export function ruleWeekendAnomaly(claims: Claim[]): RuleFlag[] {
  // Compute each provider's weekend rate
  const providerTotal = new Map<string, number>()
  const providerWeekend = new Map<string, number>()
  for (const c of claims) {
    providerTotal.set(c.provider_id, (providerTotal.get(c.provider_id) ?? 0) + 1)
    if (c.is_weekend) {
      providerWeekend.set(c.provider_id, (providerWeekend.get(c.provider_id) ?? 0) + 1)
    }
  }

  return claims.map(c => {
    if (!c.is_weekend) return flag(c.claim_id, 'weekend_anomaly', false, '')
    const hasSurgical = c.procedure_codes.some(isSurgical)
    if (!hasSurgical) return flag(c.claim_id, 'weekend_anomaly', false, '')
    const total = providerTotal.get(c.provider_id) ?? 0
    const weekend = providerWeekend.get(c.provider_id) ?? 0
    const rate = total > 0 ? weekend / total : 0
    const triggered = rate < WEEKEND_RATE_THRESHOLD
    return flag(c.claim_id, 'weekend_anomaly', triggered,
      triggered ? `Provider ${c.provider_id} weekend rate ${(rate * 100).toFixed(1)}% < 5%; surgical on weekend` : '')
  })
}

// Rule 7: Diagnosis-procedure mismatch
// For diagnoses with a known valid procedure set, flag claims that use a procedure outside that set.
export function ruleDxProcMismatch(claims: Claim[]): RuleFlag[] {
  return claims.map(c => {
    const validSet = VALID_DX_PROCS[c.diagnosis_code]
    if (!validSet) return flag(c.claim_id, 'dx_proc_mismatch', false, '')
    const invalid = c.procedure_codes.filter(p => !validSet.includes(p))
    const triggered = invalid.length > 0
    return flag(c.claim_id, 'dx_proc_mismatch', triggered,
      triggered ? `Diagnosis ${c.diagnosis_code} with invalid procedures: ${invalid.join(', ')} (valid: ${validSet.join(', ')})` : '')
  })
}

// Rule 8: Amount clustering — submitted_amount in [47500, 50000)
export function ruleAmountClustering(claims: Claim[]): RuleFlag[] {
  return claims.map(c => {
    const triggered = c.submitted_amount >= CLUSTER_LOW && c.submitted_amount < CLUSTER_HIGH
    return flag(c.claim_id, 'amount_clustering', triggered,
      triggered ? `Amount $${c.submitted_amount.toFixed(2)} in clustering range [$${CLUSTER_LOW}, $${CLUSTER_HIGH})` : '')
  })
}

// Run all 8 rules and return per-claim flag arrays
export function applyAllRules(claims: Claim[]): Map<string, RuleFlag[]> {
  const ruleFns = [
    ruleDuplicate,
    ruleRapidResubmission,
    ruleUpcoding,
    ruleUnbundling,
    rulePhantomBilling,
    ruleWeekendAnomaly,
    ruleDxProcMismatch,
    ruleAmountClustering,
  ]

  // Collect all flags per claim_id
  const claimFlags = new Map<string, RuleFlag[]>()
  for (const c of claims) claimFlags.set(c.claim_id, [])

  for (const fn of ruleFns) {
    const results = fn(claims)
    for (const f of results) {
      if (f.triggered) claimFlags.get(f.claim_id)?.push(f)
    }
  }
  return claimFlags
}
